plugin/pkg/admission/noderestriction/admission.go

```diff
@@ -655,3 +661,32 @@ func (p *Plugin) admitResourceSlice(nodeName string, a admission.Attributes) err
 
 	return nil
 }
+
+func (p *Plugin) admitCSR(nodeName string, a admission.Attributes) error {
+	// Create requests for Kubelet serving signer and Kube API server client
+	// kubelet signer with a CN that begins with "system:node:" must have a CN
+	// that is exactly the node's name.
+	// Other CSR attributes get checked in CSR validation by the signer.
+	if a.GetOperation() == admission.Create {
+		csr, ok := a.GetObject().(*certapi.CertificateSigningRequest)
+		if !ok {
+			return admission.NewForbidden(a, fmt.Errorf("unexpected type %T", a.GetObject()))
+		}
+		if csr.Spec.SignerName != certapi.KubeletServingSignerName && csr.Spec.SignerName != certapi.KubeAPIServerClientKubeletSignerName {
+			return nil
+		}
+		csrBytes := make([]byte, base64.StdEncoding.DecodedLen(len(csr.Spec.Request)))
+		_, err := base64.StdEncoding.Decode(csrBytes, csr.Spec.Request)
```

Comment:
no decoding needed here, go will have already done the base-64 decode at this point

Reply:
I'm *really* surprised the e2e tests all passed with this bogus base64-decode in place ... are none of them using kubelet-created client/serving CSRs?

Reply:
They must not be!

Reply:
This seems like a test gap that we need to address?